crates/core/src/workflows/application.rs

```diff
@@ -8,15 +8,39 @@ use holochain_core_types::{
     dna::Dna,
     error::{HcResult, HolochainError},
 };
+use log::error;
 use std::sync::Arc;
 
 pub async fn initialize(
     instance: &Instance,
-    dna: Option<Dna>,
+    maybe_dna: Option<Dna>,
     context: Arc<Context>,
 ) -> HcResult<Arc<Context>> {
     let instance_context = instance.initialize_context(context.clone());
-    let dna = dna.ok_or(HolochainError::DnaMissing)?;
+
+    // This function is called in two different cases:
+    // 1. Initializing a brand new instance
+    // 2. Loading a persistent instance from storage
+    //
+    // In the first case, we definitely need maybe_dna to be Some, because that is the DNA
+    // that will seed this instance's Nucleus.
+    // If maybe_dna is None, we expect to find a DNA in the Nucleus (=in the state) already.
+    let dna = if let Some(dna) = maybe_dna {
+        // Ok, since maybe_dna is set, we are assuming to seed a new instance.
+        // To make sure that we are not running into a weird state, we are going
+        // to check here if we really deal with a fresh state and no DNA in the Nucleus already:
+        if context.get_dna().is_some() {
+            panic!("Tried to initialize instance that already has a DNA in its Nucleus");
+        }
+        dna
+    } else {
+        // No DNA provided as parameter.
+        // This is the loading-case - we assume to find a DNA in the Nucleus state:
+        context.get_dna().ok_or_else(|| {
+            error!("No DNA provided during loading and none found in state");
```

Comment:
Should this use proper logging with the context?
```suggestion
            log_error!(context, "No DNA provided during loading and none found in state");
```